src/nbl/ui/CWindowWin32.cpp

```diff
@@ -5,6 +5,218 @@
+			wcex.hInstance = hinstance;
+			wcex.hIcon = NULL;
+			wcex.hCursor = 0; // LoadCursor(NULL, IDC_ARROW);
+			wcex.hbrBackground = (HBRUSH)(COLOR_WINDOW + 1);
+			wcex.lpszMenuName = 0;
+			wcex.lpszClassName = classname;
+			wcex.hIconSm = 0;
+
+			RegisterClassEx(&wcex);
+
+			// calculate client size
+
+			RECT clientSize;
+			clientSize.top = m_params.y;
+			clientSize.left = m_params.x;
+			clientSize.right = clientSize.left + m_params.width;
+			clientSize.bottom = clientSize.top + m_params.height;
+
+			DWORD style = WS_POPUP; // TODO why popup?
+
+			if ((m_params.flags & CWindowWin32::ECF_FULLSCREEN) == 0)
+			{
+				if ((m_params.flags & CWindowWin32::ECF_BORDERLESS) == 0)
+				{
+					style |= WS_BORDER;
+					style |= (WS_SYSMENU | WS_CAPTION);
+				}
+				// ? not sure about those below
+				style |= WS_CLIPCHILDREN;
+				style |= WS_CLIPSIBLINGS;
+			}
+			if (m_params.flags & CWindowWin32::ECF_MINIMIZED)
+			{
+				style |= WS_MINIMIZE;
+			}
+			if (m_params.flags & CWindowWin32::ECF_MAXIMIZED)
+			{
+				style |= WS_MAXIMIZE;
+			}
+			if (m_params.flags & CWindowWin32::ECF_ALWAYS_ON_TOP)
+			{
+				style |= WS_EX_TOPMOST;
+			}
+			if ((m_params.flags & CWindowWin32::ECF_HIDDEN) == 0)
+			{
+				style |= WS_VISIBLE;
+			}
+
+
+			// TODO:
+			// if (hasMouseCaptured())
+			// if (hasInputFocus())
+			// if (hasMouseFocus())
+
+			AdjustWindowRect(&clientSize, style, FALSE);
+
+			const int32_t realWidth = clientSize.right - clientSize.left;
+			const int32_t realHeight = clientSize.bottom - clientSize.top;
+
+			int32_t windowLeft = (GetSystemMetrics(SM_CXSCREEN) - realWidth) / 2;
+			int32_t windowTop = (GetSystemMetrics(SM_CYSCREEN) - realHeight) / 2;
+
+			if (windowLeft < 0)
+				windowLeft = 0;
+			if (windowTop < 0)
+				windowTop = 0;	// make sure window menus are in screen on creation
+
+			if (m_params.flags & CWindowWin32::ECF_FULLSCREEN)
+			{
+				windowLeft = 0;
+				windowTop = 0;
+			}
+
+			// create window
+
+			m_params.nativeWindow = CreateWindow(classname, __TEXT(""), style, windowLeft, windowTop,
+				realWidth, realHeight, NULL, NULL, hinstance, NULL);
+			if ((m_params.flags & CWindowWin32::ECF_HIDDEN) == 0)
+				ShowWindow(m_params.nativeWindow, SW_SHOWNORMAL);
+			UpdateWindow(m_params.nativeWindow);
+
+			// fix ugly ATI driver bugs. Thanks to ariaci
+			// TODO still needed?
+			MoveWindow(m_params.nativeWindow, windowLeft, windowTop, realWidth, realHeight, TRUE);
+		}
+		void setParams(SParams&& params) override { m_params = static_cast<SCreateWindowParams&&>(params); }
+	};
+	struct SDestroyWindowRequest : SRequest
+	{
+		SDestroyWindowRequest(CWindowWin32::native_handle_t wnd) : m_params{ wnd } {}
+		struct SDestroyWindowParams : SParams
+		{
+			CWindowWin32::native_handle_t window;
+		} m_params;
+		void process() override
+		{
+			DestroyWindow(m_params.window);
+		}
+		void setParams(SParams&& params) override { m_params = static_cast<SDestroyWindowParams&&>(params); }
+	};
```

Comment:
I have to worry you, plymorphysm wont work for requests. Request queue in IAsyncQueueDispatcher is implemented as circular buffer with storage being just an array of provided request type (`SRequest` here)
Calling `request(req_struct)` will copy just the `SRequest` part of request object (well, it won't because theres pure virtual functions, but it would want to)
tl;dr: you have to get rid of polymorphism from requests

Reply:
Oh now i see. I was wondering why it isn't implemented via dynamic polymorphism in COpenGL_Queue XD